### PR TITLE
support value attribution condition

### PR DESCRIPTION
**Pull request by yidong72**
1. support value attribute string prediction
2. support value attribute sft
3. update UI to allow user enter values.

### CODE REVIEW

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 1]
```diff
@@ -26,21 +26,34 @@
...
 
 
-def _mask_targets(target, tokenized_lens, speakers, header_len, s_ids, tokenizer, mask_role):
+def _mask_targets(target, tokenized_lens, speakers, header_len, s_ids, tokenizer, mask_role, gtype):
     cur_idx = header_len
     tgt_len = target.shape[0]
     for i, (tokenized_len, speaker, s_id) in enumerate(zip(tokenized_lens, speakers, s_ids)):
         # note, sentence piece will add extra empty token in front. s_id has that extra token too
         skip_name_len = len(tokenizer.text_to_ids(TURN_TOKEN + speaker + END_NAME_SIGNAL))
+        if (s_id[1:] == 255002).any().item():
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:48]: Can we check detokenized string value instead of a hardcode value specific to a particular tokenizer?

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml [hunk 2]
```diff
@@ -35,4 +35,9 @@ share: False  # whether create a public URL
 username: test # user name for web client
 password: test2  # password for web client
 web_port: 9889 # the port number of the web server
-chat: False # use the chat interface
\ No newline at end of file
+chat: False # use the chat interface
+chatbot_config:
+  value: False   # whether to inject the value attributes
```
Comment (MaximumEntropy) [inline examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml:40]: Can you call this steerlm_config instead?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 3]
```diff
@@ -26,21 +26,34 @@
 END_SIGNAL = "\n"
 END_NAME_SIGNAL = "\n"
 
+SCALE = 9
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:29]: Can we avoid making this a global variable?

Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:29]: It makes it hard to experimentat with different kinds of value models.

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
Few minor comments.

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml (hunk 2 above)
Comment (yidong72) [reply examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml:40]: the paper is in anonymity. don't want to mention the name of the method.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 3 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:29]: removed this. only use the value string

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 1 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:48]: done. but it is still associated with sentence piece as this tokenizer added an extra token in front. Need to test other tokenizers.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 4]
```diff
@@ -65,6 +78,13 @@ def cannonical_form_formater(cannoical_form):
     return f'<extra_id_2>{cannoical_form}\n'
 
 
+def response_value_formater(label):
+    if isinstance(label, str):
+        return '<extra_id_2>' + label + '\n'
+    else:
+        raise ValueError(f'Unknown label type {type(label)}')
```
Comment (Zhilin123) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:85]: Maybe more informative error msg (e.g. please have label in str format instead)

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 5]
```diff
@@ -29,18 +29,31 @@
...
         # note, sentence piece will add extra empty token in front. s_id has that extra token too
         skip_name_len = len(tokenizer.text_to_ids(TURN_TOKEN + speaker + END_NAME_SIGNAL))
+        if extra_id_2_id is None:
+            raise ValueError("extra_id_2 is not in the vocabulary")
+        if (s_id[1:] == extra_id_2_id).any().item():
+            if gtype == 'VALUE_TO_TEXT':
+                # if contains the token <extra_id_2>
+                assert skip_name_len - 1 == torch.where((s_id[1:] == extra_id_2_id))[0].item()
+                # find new line token id 14
+                more_skip_len = torch.where((s_id[skip_name_len:] == 14))[0][0].item()
```
Comment (Zhilin123) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:53]: Similar to extra_id_2_id, maybe don't set the new line id explicitly (in case tokenizer changes in the future)

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 6]
```diff
@@ -96,6 +116,26 @@ def _add_speaker_and_signal(header, source, mask_role, gtype):
                 + sentence["value"]
                 + END_SIGNAL
             )
+        elif gtype == "VALUE_TO_TEXT":
+            sentence["value"] = (
+                BEGIN_SIGNAL
+                + role_token
+                + sentence_from
+                + END_NAME_SIGNAL
+                + (response_value_formater(sentence['label']) if 'label' in sentence else '')
```
Comment (Zhilin123) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:125]: Can you leave a comment here indicating that loss is not calculated on `(response_value_formater(sentence['label']) if 'label' in sentence else '')` ? Don't think it's very straightforward here.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 7]
```diff
@@ -96,6 +116,26 @@ def _add_speaker_and_signal(header, source, mask_role, gtype):
...
+                BEGIN_SIGNAL
+                + role_token
+                + sentence_from
+                + END_NAME_SIGNAL
+                + sentence["value"]
+                + END_SIGNAL
+                + (response_value_formater(sentence['label']) if 'label' in sentence else '')
+            )
         else:
             raise ValueError(f"source type {gtype} not supported")
```
Comment (Zhilin123) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:140]: Maybe put here as the list of gtype accepted?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 8]
```diff
@@ -157,14 +196,21 @@ def _build_samples_mapping(self):
         assert hasattr(self.tokenizer, "vocab"), "tokenizer should have vocab property, not supported"
         assert '<extra_id_0>' in self.tokenizer.vocab, "<extra_id_0> not in the tokenizer vocab. not supported"
         assert '<extra_id_1>' in self.tokenizer.vocab, "<extra_id_1> not in the tokenizer vocab. not supported"
+        # calcuilate <extra_id_2> id value
+        if '<extra_id_2>' in self.tokenizer.vocab:
+            ids_1 = self.tokenizer.text_to_ids('<extra_id_1><extra_id_2>')
```
Comment (Zhilin123) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:201]: is this a typo?

Comment (Zhilin123) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:201]: Not sure why we can't get the text_to_ids('<extra_id_2>') directly. This looks pretty hacky

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 9]
```diff
@@ -29,18 +29,31 @@
 SYSTEM_TOKEN = "<extra_id_0>System\n"
 TURN_TOKEN = "<extra_id_1>"
 
-GUARD_RAIL_INSTRUCTION = {
+TYPE_INSTRUCTION = {
     "TEXT_TO_CANONICAL_FORM": "Given a dialogue, for each turn you need to generate a short summary called a canonical form. Generate the canonical form for the last turn in the dialogue.",
```
Comment (Zhilin123) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:33]: Are we still going to make use of canonical_forms? Otherwise, may remove it? I think in general perhaps having fewer options is better, otherwise we might risk confusing others

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 10]
```diff
@@ -29,18 +29,31 @@
...
     "CANONICAL_FORM_TO_TEXT": "Given a dialogue, for each turn we also have a short summary called a canonical form. Generate the canonical form given the last turn message and canonical form. Then generate the message.",
+    'TEXT_TO_VALUE': "",
+    'VALUE_TO_TEXT': '',
 }
 
 
-def _mask_targets(target, tokenized_lens, speakers, header_len, s_ids, tokenizer, mask_role):
+def _mask_targets(target, tokenized_lens, speakers, header_len, s_ids, tokenizer, mask_role, gtype, extra_id_2_id):
```
Comment (Zhilin123) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:40]: Can you write a description on what the code is doing using an example such as (also include the control tokens)

e.g. 

(user utterance1)  (bot utterance1 )  (user utterance2) (bot utterance)
XXXXXXXXXX       YYYY                     ZZZZZZZZZZZZ   AAAA → loss on YYYY and AAAA together

as well as what each arg does?

This function is pretty hard to read without it.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 5 above)
Comment (Zhilin123) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:53]: if possible, can we call it extra_id_2_token_id ? , the name as it is now might confuse people

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 1 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py]: Okay. I fixed the code a bit and tested on the huggingface tokenizer and confirm it is working.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 4 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:99]: added

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 5 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py]: fixed

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 8 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:232]: this is to handle the sentencepiece tokenizer which adds a special token in front. A hacky way I agree.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 9 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:34]: yes we can remove canonical forms.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 7 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:154]: done

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 6 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:139]: it is not about loss calculation but to construct the conversation according to dialogue templates.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 10 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py]: added doc

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 8 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:228]: Btw this check is super slow since `self.tokenizer.vocab` is a list not a dictionary/set. It is actually faster to do `len(self.tokenizer.text_to_ids('<extra_id_2>')) == 1`.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 6 above)
Comment (Zhilin123) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:133]: ok think this was addressed elsewhere

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 8 above)
Comment (Zhilin123) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:228]: Hmm, nit but isn't it just gonna be `self.extra_id_2_id = self.tokenizer.text_to_ids('<extra_id_2>')[-1]`

**Review by Zhilin123 — DISMISSED** [review]
Generally looks good, with some remaining minor code style issues but we can address in future PR since code freeze is today.

**Review by MaximumEntropy — APPROVED** [review]
